README.md

```diff
@@ -8,45 +8,56 @@
 
 <img src="https://user-images.githubusercontent.com/5394217/70820564-06b06e00-1dea-11ea-9680-27f661ca2a58.png" alt="Output sample" width="600"/>
 
-### Motivation
+## Motivation [↑](#xrefcheck)
 
-As the project evolves, links in documentation have a tendency to get broken. This is usually because of:
-1. File movements;
-2. Markdown header renames;
-3. Outer sites ceasing their existence.
+As a project evolves, links in markdown documentation have a tendency to become broken. This is usually because:
+1. A file has been moved;
+2. A markdown header has been renamed;
+3. An external site has ceased to exist.
 
 This tool will help you to keep references in order.
+You can run `xrefcheck` continuously in your CI pipeline,
+and it will let you know when it finds a broken link.
 
-### Aims
+## Aims [↑](#xrefcheck)
 
 Comparing to alternative solutions, this tool tries to achieve the following points:
 
 * Quickness - local references are verified instantly even for moderately-sized repositories.
 * Easy setup - no extra actions required, just run the tool in the repository root.
-Both relative and absolute local links are supported out of the box.
 * Conservative verifier allows using this tool in CI, no false positives (e.g. on sites which require authentication) should be reported.
 
-### A comparison with other solutions
+## A comparison with other solutions [↑](#xrefcheck)
 
-* [linky](https://github.com/mattias-p/linky) - a well-configurable verifier written in Rust, scans one specified file at a time and works good in pair with system utilities like `find`.
+* [linky](https://github.com/mattias-p/linky) - a well-configurable verifier written in Rust, scans one specified file at a time and works well with system utilities like `find`.
   This tool requires some configuring before it can be applied to a repository or added to CI.
 * [awesome_bot](https://github.com/dkhamsing/awesome_bot) - a solution written in Ruby that can be easily included in CI or integrated into GitHub.
   Its features include duplicated URLs detection, specifying allowed HTTP error codes and reporting generation.
   At the moment of writing, it scans only external references and checking anchors is not possible.
-* [remark-validate-links](https://github.com/remarkjs/remark-validate-links) and [remark-lint-no-dead-urls](https://github.com/davidtheclark/remark-lint-no-dead-urls) - highly configurable Javascript solution for checking local and remote links resp.
+* [remark-validate-links](https://github.com/remarkjs/remark-validate-links) and [remark-lint-no-dead-urls](https://github.com/davidtheclark/remark-lint-no-dead-urls) - highly configurable JavaScript solution for checking local and external links respectively.
   It is able to check multiple repositores at once if they are gathered in one folder.
-  Being written on JavaScript, it is fairly slow on large repositories.
+  Being written in JavaScript, it is fairly slow on large repositories.
```

Comment:
Yep, that was my personal observation at that moment.

However given that we are planning a public release, we may need to verify this again :thinking:

Reply:
I see. Would you mind if we delete this particular sentence? It's just that, even if we conclude that the application's slow, I don't think we have enough confidence to assert that it's slow _because_ of being written in js :/ Also, "slow" can be subjective, and I think this section should be more objective when describing other people's projects.

Reply:
I agree that making such statements does not sound too proper :thinking: 

On the other hand, if we are just left with the mention of this alternative solution without mentioning any weak point, then it's not clear why another tool (xrefcheck) was necessary.

From my experience, at least one of those JavaScript-based tools was _noticeably_ slow. I had an impression that it didn't do any parallelization (thus the claim about JS being the reason, but such a claim is really invalid), and it took a dozen seconds to check a small-to-middle-sized repository that contained only local (!) links.

Given that, maybe we could form more objective statements here instead?

Reply:
So, I tried using `remark` on the xrefcheck repo, here are the steps to reproduce my experiment:

```
$ npm init
$ npm install remark-cli remark-lint-no-dead-urls remark-validate-links --save-dev
```

Edit `package.json` and this:

```
  "remarkConfig": {
    "plugins": [
      "remark-lint-no-dead-urls",
      "remark-validate-links"
    ]
  }
```

Run it and time it:

```
$ time npm run env -- remark . --frail --ignore-pattern 'tests/markdowns/**/*' --ignore-pattern 'tests/golden/**/*'
```

It seems to take about 8s on average to verify the xrefcheck repo.

The `xrefcheck` tools takes about 6~8s, so it's not a huge difference :/

However, I did notice that it does not handle "429 Too Many Requests" - so eventually links to github.com start failing and the tool reports false positives.

I replaced our claim about it being slow with a new claim. And also added "Resilience" as one of xrefcheck's aims.

Please have a look at the latest fixup and let me know what you think ^^

Reply:
LGTM (learned new English words, e.g. resilience). As I understand there was no waiting for 429's during speed comparison (since default waiting interval is 30s) so it was fair test . Maybe one could compare speed on a repository with many local links since your test looks like competition of network libraries.

Reply:
Created a file with 240 local links, re-ran the experiment and this time also ignored `CHANGES.md` (where most of our external links are). xrefcheck took 1.5~2s, remark took 2.5~3s. So not a huge difference here either :/

Reply:
Thanks for this exploration!

I also wanted to see a check for local links only, but it turns out to be not faster...

So yeah, the rewrite of the claim looks good :+1: 

> xrefcheck took 1.5-2s

This is quite suspicious though, to me it looks like there is nothing that forced the time to be more than a portion of second.

Initially, I wanted to make xrefcheck instant when checking local links (so that several seconds time shown by other tools seemed really long). Most user's mistakes come from local links (external links are usually copy-pasted), and a shorter time would facilitate documentation writing a lot.

Created #219 to investigate this later.

If after optimizations `xrefcheck`  takes like 0.4s on your example, how do you think, would it be correct to include the mention of xrecheck being faster? Not in vague and incorrect terms as before, but referring to our experiment showing that `remark` takes X seconds and `xrefcheck` takes Y?

Reply:
> This is quite suspicious though, to me it looks like there is nothing that forced the time to be more than a portion of second.

Oh I just remember I compiled xrefcheck with no optimizations (`stack install --fast`) ^^ Just tried again with `stack install`, but got very similar results (1.3~1.5secs) :/

Another interesting data point: I re-ran the experiment, this time with `--mode local-only`. Like before, I added [this file with 242 local links](https://gist.github.com/dcastro/3a7f38d6a0a724e72bc6e01cc6dfca33) to the xrefcheck repo (don't forget `git add a.md`), and then ran:

```
$ time xrefcheck --ignore 'tests/markdowns/**/*' --ignore 'tests/golden/**/*' --ignore CHANGES.md --mode local-only
1.05s user 0.28s system 370% cpu 0.358 total
```

0.3 secs :open_mouth: 

Even the 0.75 seconds you observed in morley in #219 is in my opinion _really_ fast :sweat_smile: 

Regarding #219: I'm tempted to say that xrefcheck is pretty much as fast as it can be.
I _suspect_ the primary bottleneck is in the network when checking external links (and is the reason for the 1.5-2 secs I saw earlier) and the secondary bottleneck in disk reads, and we can't do anything about either of those :/  
Any performance gains we can squeeze out will (I think) be very marginal, and unnoticeable by a real world user.

-----

> If after optimizations xrefcheck takes like 0.4s on your example, how do you think, would it be correct to include the mention of xrecheck being faster? Not in vague and incorrect terms as before, but referring to our experiment showing that remark takes X seconds and xrefcheck takes Y?

I'd be okay with that, but if we decide to do that, then I think it would be fair to construct a proper reproducible benchmark and compare xrefcheck against the other alternatives we mentioned in the readme (not just `remark`), and then publish a table with the results.

Reply:
I agree that likely I/O takes the most part of the time and we don't be able to introduce any significant performance improvement, unless there is some place in the verification logic that appears to be very non optimal. 
If I fail to find such places, it would be reasonable to abort the issue.

> then I think it would be fair to construct a proper reproducible benchmark and compare xrefcheck against the other alternatives we mentioned in the readme

Really, agreed.